### PR TITLE
refactor (human-robot-interface): various enhancements

### LINKED ISSUE #178: Wake word (RAI_ASR)
**Is your feature request related to a problem? Please describe.**
RAI_ASR uses silero-vad for voice activity detection. Due to this design, there is a considerable number of false positive detections, degrading the overall experience. 

**Describe the solution you'd like**
RAI_ASR should implement an optional wake word. This way every interaction would have to start with the same word eg Ray (reɪ) which will result in reducing the false positives. 

**Describe alternatives you've considered**

**Additional context**
- https://github.com/dscripka/openWakeWord
- https://github.com/KoljaB/RealtimeTTS

### PR DESCRIPTION
**Pull request by maciejmajek**
## Purpose

RAI hri (Human Robot Interface) needed enhancements to handle various hardware configurations, including microphones with different sampling frequency. 

## Proposed Changes
- fix(rai_asr): asr stream falling behind due to MultiThreadedExecutor lagging (?)
- feat(rai_asr): allowed input devices with various sampling rates via resampling to 16KHz
- feat(rai_asr): wake word using OpenWakeWord
- refactor(rai_hmi): use model defined in config.toml
- refacotr(rai_bringup): reparemetrize asr launch
- refactor: use vad model instead of vad iterator 

## Testing
```
ros2 launch rai_bringup hri.launch.py keep_speaker_busy:=false recording_device:=7 silence_grace_period:=0.5 asr_vendor:=whisper # local solution, use openai for cloud based
```

### CODE REVIEW

File: src/rai_asr/rai_asr/asr_node.py [hunk 1]
```diff
@@ -177,51 +283,45 @@ def capture_sound(self):
             if asr_lock or self.hmi_lock or self.tts_lock:
                 continue
 
-            speech_dict = self.vad_iterator(audio_data, return_seconds=True)
-            if speech_dict:
+            self.audio_buffer.append(audio_data)
```
Comment (wsiekierska) [inline src/rai_asr/rai_asr/asr_node.py]:
```suggestion
```
We should only append to buffer when we're recording

File: src/rai_asr/rai_asr/asr_node.py [hunk 2]
```diff
@@ -177,51 +283,45 @@ def capture_sound(self):
...
+            if self.should_listen(audio_data):
+                self.silence_start_time = datetime.now()
                 if not self.is_recording:
                     self.start_recording()
-                self.audio_buffer.append(audio_data)
-                self.silence_start_time = None
-                if "end" in speech_dict.keys():
-                    self.silence_start_time = datetime.now()
+                    self.reset_buffer()
```
Comment (wsiekierska) [inline src/rai_asr/rai_asr/asr_node.py]:
```suggestion
                    self.reset_buffer()
                    self.audio_buffer.append(audio_data)
```

File: src/rai_asr/rai_asr/asr_node.py [hunk 3]
```diff
@@ -177,51 +283,45 @@ def capture_sound(self):
...
             elif self.is_recording:
```
Comment (wsiekierska) [inline src/rai_asr/rai_asr/asr_node.py]:
```suggestion
            elif self.is_recording:
                self.audio_buffer.append(audio_data)
```

File: src/rai_asr/rai_asr/asr_node.py [hunk 4]
```diff
@@ -177,51 +283,45 @@ def capture_sound(self):
...
-                self.audio_buffer.append(audio_data)
-                if self.silence_start_time is not None:
-                    if datetime.now() - self.silence_start_time > self.grace_period:
-                        self.stop_recording_and_transcribe()
+                if datetime.now() - self.silence_start_time > self.grace_period:
+                    self.stop_recording_and_transcribe()
+                    self.reset_buffer()
```
Comment (wsiekierska) [inline src/rai_asr/rai_asr/asr_node.py]:
```suggestion
                    self.reset_buffer()
                    self.oww_model.reset()
```

File: src/rai_tts/rai_tts/tts_node.py [hunk 5]
```diff
@@ -53,7 +53,7 @@ def __init__(self):
         self.job_id: int = 0
         self.queued_job_id = 0
         self.tts_client = self._initialize_client()
-        self.create_timer(0.01, self.status_callback)
+        self.create_timer(0.25, self.status_callback)
```
Comment (boczekbartek) [inline src/rai_tts/rai_tts/tts_node.py]: How about declaring this as a `ros2 parameter`? 
It's hard to know the meaning of this param at the first glance

File: src/rai_hmi/rai_hmi/hmi_node.py [hunk 6]
```diff
@@ -82,7 +82,7 @@ def __init__(self):
         )
 
         self.create_timer(
-            0.01, self.status_callback, callback_group=self.callback_group
+            0.25, self.status_callback, callback_group=self.callback_group
```
Comment (boczekbartek) [inline src/rai_hmi/rai_hmi/hmi_node.py]: "magic number" - please change to `ros2 paramter`

File: src/rai_asr/rai_asr/asr_node.py [hunk 7]
```diff
@@ -135,11 +226,21 @@ def _setup_publishers_and_subscribers(self):
...
+            from rai_asr.asr_clients import OpenAIWhisper
+
+            self.model = OpenAIWhisper(
+                self.model_name, DEFAULT_SAMPLING_RATE, self.language
+            )
+        elif self.model_vendor == "whisper":
+            from rai_asr.asr_clients import LocalWhisper
+
+            self.model = LocalWhisper(
+                self.model_name, DEFAULT_SAMPLING_RATE, self.language
```
Comment (boczekbartek) [inline src/rai_asr/rai_asr/asr_node.py]: `DEFAULT` part in `DEFAULT_SAMPLING_RATE` suggests that this value can be changed, but it doesn't. I would name it `SAMPLING_RATE` or allow parametrization if that's the case

File: src/rai_asr/rai_asr/asr_node.py [hunk 8]
```diff
@@ -153,80 +254,120 @@ def hmi_status_callback(self, msg: String):
...
-                self.audio_buffer.append(audio_data)
-                if self.silence_start_time is not None:
-                    if datetime.now() - self.silence_start_time > self.grace_period:
-                        self.stop_recording_and_transcribe()
+    def reset_buffer(self):
+        self.audio_buffer.clear()
 
     def start_recording(self):
-        self.get_logger().info("Recording...")
+        self.get_logger().info("Recording...")  # type: ignore
```
Comment (boczekbartek) [inline src/rai_asr/rai_asr/asr_node.py:342]: Can't see type conflict here, did it break `flake8` for you?
```suggestion
        self.get_logger().info("Recording...") 
```

File: src/rai_asr/rai_asr/asr_node.py [hunk 9]
```diff
@@ -153,80 +254,120 @@ def hmi_status_callback(self, msg: String):
...
         self.publish_status("recording")
         self.is_recording = True
-        self.audio_buffer = []
-        self.silence_start_time = None
 
     def stop_recording_and_transcribe(self):
-        self.get_logger().info("Stopped recording. Transcribing...")
+        self.get_logger().info("Stopped recording. Transcribing...")  # type: ignore
```
Comment (boczekbartek) [inline src/rai_asr/rai_asr/asr_node.py:347]: Can't see type conflict here, did it break `flake8` for you?
```suggestion
        self.get_logger().info("Stopped recording. Transcribing...")
```

File: src/rai_asr/rai_asr/asr_node.py [hunk 10]
```diff
@@ -153,80 +254,120 @@ def hmi_status_callback(self, msg: String):
...
         self.is_recording = False
         self.publish_status("transcribing")
         self.transcribe_audio()
         self.publish_status("waiting")
+        self.get_logger().info("Done transcribing.")  # type: ignore
```
Comment (boczekbartek) [inline src/rai_asr/rai_asr/asr_node.py:352]: Can't see type conflict here, did it break `flake8` for you?
```suggestion
        self.get_logger().info("Done transcribing.") 
```

**Review by boczekbartek — CHANGES_REQUESTED** [review]
@maciejmajek I left some minor comments about the code.
Also, shouldn't [docs/human_robot_interface/voice_interface.md](https://github.com/RobotecAI/rai/blob/41b86ced428d04c35bc6bbd7d7c18c727effe1b1/docs/human_robot_interface/voice_interface.md) be updated (for e.g. about cloud / local `Whisper` setup)

File: src/rai_asr/rai_asr/asr_node.py (hunk 8 above)
Comment (maciejmajek) [reply src/rai_asr/rai_asr/asr_node.py:342]: It's breaking my pyright
```
Type of "info" is partially unknown
  Type of "info" is "(message: Unknown, **kwargs: Unknown) -> bool"Pylance[reportUnknownMemberType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportUnknownMemberType)
  ```

File: src/rai_asr/rai_asr/asr_node.py (hunk 7 above)
Comment (maciejmajek) [reply src/rai_asr/rai_asr/asr_node.py]: Renamed to `VAD_SAMPLING_RATE`

**Review by boczekbartek — APPROVED** [review]
Very good work!

**Pull request merged**